Hygienic import of rewrite rules

andreasabel opened the pull request:
Since the imported signature contains all rewrite rules we visited during loading various modules (see #4272), we need to filter out those that are currently not (transitively) imported.  To this end:
1. Label each rewrite rule with its `TopLevelModuleName`.
2. Maintain in the state a set of transitively imported modules.
3. When considering a rewrite rule, check whether its `TopLevelModuleName` is in the latter set.
  
Closes #4343.
Closes #7933.
Closes #7935. 
Rendering: https://agda--7934.org.readthedocs.build/en/7934/language/rewriting.html#importing-rewrite-rules

andreasabel commented:
@jespercockx : Can you test this PR on your REWRITE examples, also the confluence examples and check whether introduces any regressions?
I think we are lacking a complicated module structure for rewriting in our testsuites.

ecavallo commented:
@andreasabel asked me if I had any project that uses rewrite rules. Here is one: https://github.com/ecavallo/equivariant-cartesian

andreasabel commented:
> @andreasabel asked me if I had any project that uses rewrite rules. Here is one: https://github.com/ecavallo/equivariant-cartesian

Thanks, seems that my PR breaks this.

On a fresh checkout, I get for `agda summary.agda`:
```
Checking summary (/Users/abel/bib/authors/CavalloEvan/equivariant-cartesian/summary.agda).
 Checking basic (/Users/abel/bib/authors/CavalloEvan/equivariant-cartesian/basic.agda).
...
   Checking tiny.preserves-coproduct (/Users/abel/bib/authors/CavalloEvan/equivariant-cartesian/tiny/preserves-coproduct.agda).
    Checking tiny.basic (/Users/abel/bib/authors/CavalloEvan/equivariant-cartesian/tiny/basic.agda).
/Users/abel/bib/authors/CavalloEvan/equivariant-cartesian/tiny/preserves-coproduct.agda:41.28-58: error: [UnequalTerms]
transposeLeft (back ∘ inl) f != inl f of type
(_Γ_25 ^ S) ⊎ (_B_30 ^ S)
when checking that the inferred type of an application
  (transposeLeft back ∘ (inl `^ S)) (λ z → f z) ≡
  transposeLeft (back ∘ inl) (λ z → f z)
matches the expected type
  transposeLeft back (forward (inl f)) ≡ inl f
```
So I guess my PR loses some rewrite rules...

andreasabel committed:
Test for #4343: agda-2.7.0.1 --ignore-interfaces Main.agda (fails)

This test succeeds when rewrite rules are imported hygienically.
The rules of A1 must not be present when checking B2 (which imports B1).
The rules of B1 must not be present when checking A2 (which imports A1).
Main needs the rules of both A2 and B2.

andreasabel committed:
Testcase for #4343 and #7933

andreasabel committed:
Debug printing imports for #4343

andreasabel committed:
Refactor for #4343: add TopLevelModuleName to RewriteRule

Prepares for scope checking rewrite rules so that only rewrite rules
from (transitively) imported modules are usable.

andreasabel committed:
Refactor for #4343: maintain set of transitively imported modules in state

andreasabel committed:
Fix #4343 and #7933: only use rewrite rules from imported modules

Rewrite rules that happen to be in stImports from modules that we
visited but that are currently not imported are filtered out now by
checking whether their TopLevelModule is (transitively) imported.

Closes #4343.
Closes #7933.

andreasabel committed:
Debug printing transitive imports for #4343

andreasabel committed:
Refining a debug message for the confluence checker

andreasabel committed:
Restore behavior of test case for issue #4333

andreasabel commented:
I replaced the PR by a new fix and now it passes @ecavallo 's development. Yay!

andreasabel commented on test/interaction/Issue4333.out:
```diff
@@ -2,11 +2,3 @@ Checking Issue4333.M (Issue4333/M.agda).
-Possible fix: add a rule to rewrite Issue4333.M.a₁' to
-Issue4333.M.a₀'
-when scope checking the declaration
-open import Issue4333.N1
```
@jespercockx : This testcase no longer reports a confluence problem.
This seems in alignment with the description of the test case in https://github.com/agda/agda/blob/ebfd636f2da231a995cb8618afade3f7236434b9/test/interaction/Issue4333.hs#L11-L14

Now indeed N0 and N1 are checked independently what the rewrite rules in scope concerns.

However, I still think the confluence checker should complain for module N.  Both N0 and N1 are confluent by themselves, but their union introduces a critical pair (a0, a1).  Even though N does not introduce any rewrite rules, shouldn't it rerun the confluence checker once it merges N1 into the signature?

It seems to do so, finding a critical pair, but this does not cause an error (anymore):
``` 
agda -vimport.iface.confluence:20 -vrewriting.confluence:10 --ignore-interfaces N.agda
Checking Issue4333.N (/Users/abela/agda/test/interaction/Issue4333/N.agda).
 Checking Issue4333.M (/Users/abela/agda/test/interaction/Issue4333/M.agda).
  Checking confluence of imported rewrite rules
 Checking Issue4333.N0 (/Users/abela/agda/test/interaction/Issue4333/N0.agda).
(re)checking triangle property for rule p₀
Checking confluence of rule p₀
  Checking confluence of imported rewrite rules
Checking confluence of rule Issue4333.M.p₀
 Checking Issue4333.N1 (/Users/abela/agda/test/interaction/Issue4333/N1.agda).
(re)checking triangle property for rule p₁
Checking confluence of rule p₁
  Checking confluence of imported rewrite rules
(re)checking triangle property for rule Issue4333.M.p₀
Checking confluence of rule Issue4333.M.p₁
Found critical pair: 
  Issue4333.M.a
 ---> 
  Issue4333.M.a₁'
 =?= 
  Issue4333.M.a₀'
 : 
  Issue4333.M.A
```

jespercockx replied:
This test case is weird, it looks like `Issue4333.agda` itself is empty so it makes sense for there to be no warning. Perhaps we should move the contents of `Issue4333/N.agda` (which contains the actual imports) to `Issue4333.agda`.

jespercockx replied:
Also note that this test case has a custom test script in https://github.com/agda/agda/blob/master/test/interaction/Issue4333.hs so that's perhaps why it exhibits unexpected behavior here.

andreasabel replied:
The test script ensures that the `.agdai` files for the imported modules are built in isolation.
This PR should allow us to drop this workaround. 

However, the test script also wants `N` to fail.  _With this PR, `N` does not fail anymore._
So I suppose there is a mistake in this comment.
> -- Upon proper handling of that issue, this test case 
> -- can be moved to test/Succeed (with N as main file). 

The author @sattlerc probably meant to write `test/Fail`.

This means that this PR introduces a regression in the confluence checker.
If we are ok with this, giving confluence checking a lower prio than hygienic rewrite imports, I can merge this PR nevertheless.

jespercockx replied:
I still don't get why your PR makes N no longer report a confluence issue though. You didn't revert the fix of #4333, so why should a change in how imports of rewrite rules are processed change how the confluence checker treats them? I'd prefer to actually know *why* this breaks before merging.

andreasabel replied:
It might be connected to the `sortRulesForSymbol` whose behavior changed when `getRewriteRulesFor` only returned the rewrite rules in scope.  In the last commit, I restored its old (unhygienic) behavior.
I find `sortRulesForSymbol` questionable in the presence of imported modules, more information in #7969.
Anyway the regression is fixed...

sattlerc replied:
> The author @sattlerc probably meant to write test/Fail.

That seems likely.

andreasabel commented on test/BuildSucceed/Issue4343/Base.agda:
```diff
@@ -0,0 +1,16 @@
+
+not : Bool → Bool
+not true  = b
+not false = a
```
The `not` is superfluous.

andreasabel commented on src/full/Agda/TypeChecking/Monad/Base.hs:
```diff
@@ -2313,6 +2322,8 @@ data RewriteRule = RewriteRule
   , rewType    :: Type       -- ^ @Γ ⊢ t@.
   , rewFromClause :: Bool    -- ^ Was this rewrite rule created from a clause in the definition of the function?
+  , rewTopModule  :: TopLevelModuleName
+      -- ^ In which file is this rewrite rule defined?
```
Maybe add to this comment the purpose of this field: to filter rewrite rules from the imported signature.

jespercockx commented:
> Can you test this PR on your REWRITE examples, also the confluence examples and check whether introduces any regressions?

I don't have any multi-module rewrite examples.

jespercockx commented on test/BuildSucceed/Issue4343/Main.agda-lib:
```diff
@@ -0,0 +1,2 @@
+include: .
+flags: --rewriting
```
There's a missing newline here at EOF (maybe we should update `fix-whitespace` to also take into account `.agda-lib` files?)

andreasabel replied:
Ok, did this.

andreasabel committed:
Fix #7935: document scoping of rewrite rules

andreasabel committed:
Add *.agda-lib to fix-whitespace.yaml

andreasabel commented:
Added documentation for importing rewrite rules (closing #7935).

andreasabel commented:
@jespercockx : I am planning to merge this today and make a new RC; this is currently the last open release-blocker.
So if you want to evaluate it further, please act now.

Because of e.g. #7962 I feel the confluence checker is very much WIP and probably not used a lot, so potentially causing some regression there is a risk we can take.

jespercockx commented:
Thank you for looking into the regression and fixing it, I'm happy with this PR being merged as it is!

(I agree with your perception that the confluence checker is not being used much.)
